src/transformStateWithClones.js

```diff
@@ -1,13 +1,42 @@
 'use strict';
 
 /**
- * @param {Object} state
+ * @param {Object} stateCopy
  * @param {Object[]} actions
  *
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultsOfActions = [];
+  let stateCopy = { ...state };
+
+  for (const objectFromActionsArray of actions) {
+    switch (objectFromActionsArray.type) {
+      case 'addProperties':
+
+        Object.assign(stateCopy, objectFromActionsArray.extraData);
+
+        resultsOfActions.push({ ...stateCopy });
+        break;
+      case 'removeProperties':
+        const keyNamesArray = objectFromActionsArray.keysToRemove;
+
+        for (const keyName of keyNamesArray) {
+          delete stateCopy[keyName];
+        }
+        resultsOfActions.push(Object.assign({}, stateCopy));
+        break;
+
+      case 'clear':
+        const empty = {};
+
+        stateCopy = empty;
+        resultsOfActions.push({ ...stateCopy });
```

Comment:
You push a copy of stateCopy in all cases. Just do it once afterwards
![image](https://github.com/mate-academy/js_stateful-clones/assets/96409563/7591c3d3-1346-4320-9dfe-a71a51caa160)